triggers/daml/MatchingEngine.daml

```diff
@@ -0,0 +1,122 @@
+-- Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+module MatchingEngine where
+
+import DA.Action
+import DA.Foldable hiding (elem, null, length)
+import DA.List
+import Daml.Trigger
+
+import qualified Marketplace.Trading.Order as Order
+import qualified Marketplace.Trading.Service as Service
+import qualified Marketplace.Trading.Matching as Matching
+
+import Utils
+
+type CurrentOrderId = Int
+
+handleMatching : Trigger CurrentOrderId
+handleMatching = Trigger
+  { initialize = return 0
+  , updateState = \_ -> pure ()
+  , rule = handleMatchingRule
+  , registeredTemplates = RegisteredTemplates [ registeredTemplate @Service.CreateOrderRequest
+                                              , registeredTemplate @Service.CancelOrderRequest
+                                              , registeredTemplate @Order.T ]
+  , heartbeat = None
+  }
+
+handleMatchingRule : Party -> TriggerA CurrentOrderId ()
+handleMatchingRule party = do
+  debug "Running matching rule..."
+
+  -- Acknowledge all 'Order.Request' and update current ID
+  orderRequests <- query @Service.CreateOrderRequest
+  forA_ orderRequests \(cid,_) -> do
+    currentOrderId <- get
+    emitExerciseCmd cid Service.AcknowledgeRequest with providerOrderId = show currentOrderId
+    debug "Acknowledging order"
+    modify (+1)
+
+  -- Acknowledge all 'Order.CancelRequest'
+  orderCancelRequests <- query @Service.CancelOrderRequest
+  forA_ orderCancelRequests \(cid,_) -> emitExerciseCmd cid Service.AcknowledgeCancel
+                                     >> debug "Acknowledging order cancel"
+
+  -- Check for matches on all 'Order'
+  orders <- query @Order.T
+  forA_ orders (handleOrder party orders)
+
+-- |Check for crossing orders. If found, fill both orders. Matches by price.
+handleOrder : Party -> [ContractPair Order.T] -> ContractPair Order.T -> TriggerA CurrentOrderId ()
```

Comment:
Checking the definition, it looks like its hardcoded to return `()`

> -- | This is the type of your trigger. `s` is the user-defined state type which
-- you can often leave at `()`.
data Trigger s = Trigger
  { initialize : TriggerInitializeA s
  -- ^ Initialize the user-defined state based on the ACS.
  , updateState : Message -> TriggerUpdateA s ()
  -- ^ Update the user-defined state based on a transaction or
  -- completion message.  It can manipulate the state with `get`, `put`,
  -- and `modify`, or query the ACS with `query`.
  **, rule : Party -> TriggerA s ()**
  -- ^ The rule defines the main logic of your trigger. It can send commands
  -- to the ledger using `emitCommands` to change the ACS.
  -- The rule depends on the following arguments:
  --
  --   * The party your trigger is running as.
  --   * The user-defined state.
  --
  -- and can retrieve other data with functions in `TriggerA`:
  --
  --   * The current state of the ACS.
  --   * The current time (UTC in wallclock mode, Unix epoch in static mode)
  --   * The commands in flight.
  , registeredTemplates : RegisteredTemplates
  -- ^ The templates the trigger will receive events for.
  , heartbeat : Optional RelTime
  -- ^ Send a heartbeat message at the given interval.
  }